### ISSUE TITLE
Keep track of solved cells

### ISSUE THREAD

**Issue by mikemcbride** [mikemcbride/vue-sudoku#11]
We perform an operation each time we need to get all solved cells where we loop through the grid and check if the cell is solved. We would probably be better served by keeping track of unsolved cells and solved cells in two separate `Set`s and managing that list when we mark a cell as solved. That would cut down on time when building a list of solved and unsolved cells used internally for other calculations.

**Comment by mikemcbride** [comment]
I just found this repo that would probably make quite a difference given the number of Array manipulations we're doing:

https://github.com/lukeed/arr

**Comment by mikemcbride** [comment]
That library didn't really make much of an impact unfortunately. I think there are other avenues we can explore to avoid having to check available cell values multiple times. For instance, when fetching solved cells, we map over the grid and filter out unsolved items. We could keep track of solved cells by their coordinates in a `Set`.

We also are spending a fairly significant portion of time doing array loops and operations around checking related cell values (cells in the same row, column, or square). We can probably introduce some optimizations around these places.

**Comment by mikemcbride** [comment]
I've been working on some array optimizations with positive results, but nothing overwhelming by any means.

Keeping track of the unsolved cells still feels like it could be impactful, but I tried a couple of approaches and they didn't really yield good results. I might be coming at it the wrong way.

Also feels like optimizing time in specific methods could be good. This is where I've seen the positive results on my local branch, is in optimizing some of these methods that are called the most often. Here's a snapshot of method invocation counts during a single run of the "evil" level puzzle:

![image](https://user-images.githubusercontent.com/6248612/74818321-09d14a00-52c4-11ea-87d2-e61eafd1f230.png)

Increment Calculations is just a helpful method for `+1` on calculation count so there's nothing really to do there. I think the refactor I did on the related cells calls ended up being the right thing, but maybe there's more room for improvement. I refactored that in a way that eliminated two passes over the array each time the `getRelatedCells` method is called. So it was effectively doing 3 loops over the array and I cut it down to one through some refactors. These array methods are already extremely fast and our data set is tiny - the grid is only 81 items so it's not like we're dealing with hundreds of thousands of items or anything.

I've been doing some reading on other strategies and there may be some potential to explore those avenues, but it would require some pretty hefty refactoring of the data model so I'm not super keen on going down that path right now.

Probably will have a PR soon for the new array methods and optimization stuff.

**Comment by mikemcbride** [comment]
Jotting down a thought before I call it a night. Would it be possible to save time doing lookups on related solved/unsolved cells if we took the time to make an array of related cell IDs at the initial setup? We could then iterate over these cell ids instead of building them out each time and then filtering. Then we could check if the cell is solved by doing something like:

```js
for (let cellId of cell.relatedCells) {
  if (this.solvedCells.has(cellId) {
    // the related cell is solved
  } else {
    // the related cell is unsolved
  }
}
```

It would be super easy to generate the list of related cell IDs up front. We've already got the code to do it... could reduce some calls. Worth exploring.
